### ISSUE TITLE
LSTMs: Performance overhead due to JS way worse than gain from using WebGL

### ISSUE THREAD

**Issue by rumschuettel** [tensorflow/tfjs#1213]
Hi!

We're currently using tf.js for an LSTM that's executed in the browser on a real-time input stream. We noticed that the network becomes almost unusably slow when the number of layers increases beyond 3,/the layer size increases beyond 1024.

To figure out where the bottleneck is, I performed a series of profiling runs, such as:
![image](https://user-images.githubusercontent.com/16973848/52461770-59176300-2b25-11e9-9338-e643790eaa01.png)

What's shown there is the 3-layer/1024 wide LSTM that's executed once every 200ms on average; the curious aspect about it is that the overhead from DOM (we have almost no DOM updates) and JS (the tf.js-functions make up almost all of the time) is enormous, as compared to the time actually spent within the WebGL context rendering the shaders.

This is on an i7 with the intel graphics, not a dedicated GPU.

I'd be happy to help pin down ways to reduce this overhead. Going down the call tree unveils that there's quite a significant amount of time spent outside of the WebGL context:
![image](https://user-images.githubusercontent.com/16973848/52461756-4866ed00-2b25-11e9-99f7-f17a4d6aede3.png)

I'd be happy to help figure out how to reduce this overhead if there's interest from the side of the developers; I'm also happy for an answer à la "wontfix", at least then I'll know to look elsewhere.

Thanks a lot already,
Johannes

**Comment by caisq** [comment]
@rumschuettel What version of TF.js are you using?

**Comment by rumschuettel** [comment]
I could reproduce this in 0.14.1, then updated to 0.15 where it appears
unchanged, then 1.0.0-alpha, with the same outcome.

**Comment by caisq** [comment]
@rumschuettel FYI, it is in our roadmap to improve performance and RNN performance is a part of it. Can you share with me the basic configuration of your model. Sharing the model JSON would be idea, however, if that's not possible (e.g., due to propriety concerns), telling me the config of your LSTM part would be sufficient, as long as it helps reproducing the performance problem. You mentioned `units` is 1024, how about other key properties such as `returnSequences`? How many time steps are there in your input? What's the input shape? Also, is this inference or training?

**Comment by rumschuettel** [comment]
Inference. Example model details for which the slowdown is very noticeable:
3 layers, width 512, input size 30, output size 42.
We feed a sequence of length 1 per call (as our input is incremental), and a batch size of 1.
`trainable` is false everywhere, and both `returnSequences` as well as `stateful` is true.

Would it be significantly faster to feed in sequences of, say, size 10, instead of 10 times calling predict with a sequence of length 1?

**Comment by rumschuettel** [comment]
I can confirm that calling the LSTM with a sequence of length 10 instead of 10 times calling it with a sequence of length 1 reduces the JS overhead significantly (roughly 10-15%). The time ratio is now
Javascript: 25%
DOM: 15%
Graphics: 10%
as compared to what's shown in the first screenshot.

**Comment by rumschuettel** [comment]
So the core overhead in JS (apart from the numerous nested `tf.tidy`-stuff) is to be found in `LSTMCell`, an anonymous function inside that only shows up as `o`, and `mul_`.

All of them do significant calls to the GPU; yet for each of them, the JS overhead is roughly twice the time spent on the GPU.

![image](https://user-images.githubusercontent.com/16973848/52526322-b74e5e00-2c6b-11e9-9c4d-cd51faaf5b56.png)

**Comment by caisq** [comment]
@dsmilkov When you get a chance, I'd like to chat with you about the profile picture provided above. `mul_` is from tfjs-core and `o` is probably the argument function to `tidy()`. I want to see if there is by any chance any overhead related to shader compilation here.

**Comment by rumschuettel** [comment]
please let me know if I can be of any help.

**Comment by dsmilkov** [comment]
Thanks @rumschuettel for the analysis. A few thoughts:

- Any performance optimizations are more than welcome! If you do such optimizations and send us a PR, please also share benchmark code so we can reproduce it and make sure it doesn't negatively impact other use-cases.
- It is hard for us to analyze what's going on from screenshots like this.
  - Can you send us pointers to your code?

Tips for better profiling I've observed:
  - Make a minimal reproducible example with `<script>` and vanilla js. Depend on un-minified version of tf-core (e.g. [this one for core](https://unpkg.com/@tensorflow/tfjs-core@0.15.3/dist/tf-core.js)) and similarly for layers, so there is less function renames.
  - Even better, build tfjs-core and tfjs-layers from source, with "esnext" as compilation typescript target, at which point the stack trace is much cleaner since typescript won't inject polyfills for new features like `async/await`.
  - Good to check if Chrome or another browser gives a similar profile, in which case we know this is not Firefox-only
  - Seeing lots of tidy() calls in the profile is not necessarily a problem since they are cheap.
  - Make sure to run at least one inference between starting to profile, since the very first inference is dominated by shader compilation. After that we cache.

**Comment by dronus** [comment]
Also you have to be aware that CPU profiling is somewhat flawed in respect to tensorflowjs: As WebGL is used here for major calculations, a lot of GPU jobs are created, which will be run in parallel by the graphics driver. The time those computations take however are only visible if the CPU syncs to them, eg, by reading back the result into a JS variable. So multiple large calculations may be scheduled to happen on the GPU, while the JS functions seem to excute fast, but at some point the result is read back, or some other command requires the GPU process to complete - now we see a JS call taking a long time to wait for the graphics system to complete. So times reported by gl.readPixels, gl.useProgram, gl.whatever may be in fact the time needed to do computations issued somewhere else in the tensorflow.js code. 
Thus said, it is often hard to impossible to get the run time of tensorflow primitives by the profiler. Only way to be sure is doing experiments with modified code or parameters to get an indirect clue on the cost of single operations.
